### PR TITLE
feat: Deprecate "import assertions" with a warning

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit deprecates "import assertions" proposal that has been
replaced with "import attributes".

Any time an import assertion is encountered a warning will be printed
to the terminal. This warning will be printed for both local and 
remote files (ie. user code and dependencies).

Import assertions support will be removed in Deno 2.

### CODE REVIEW

**Comment by marvinhagemeister** [comment]
Should this be a linting rule so that users can call `deno lint --fix` to update their code?

**Comment by bartlomieju** [comment]
I'm not sure if that's feasible, I don't see anything in SWC's AST that would indicate that; see https://swc-ast.vercel.app/ and paste `import data from "./data.json" assert { type: "json" };`. The produced AST contains "with" property, but there's no indication of "assert" keyword. @dsherret thoughts on this one?

**Comment by bartlomieju** [comment]
Also, @lucacasonato wants us to warn on usage of "assert" in the dependencies, which `deno lint` would not catch.

I'm currently fighting with V8 though, it appears that if V8 code cache is used then the warning is not printed - ie. only the first time you execute the warning is printed.

**Comment by bartlomieju** [comment]
Did more digging and the problem with message appearing only once is indeed the V8 code cache. So to land this PR we need to do a few things:
- ensure that V8 code cache is not saved (or purged) for modules that raised this warning
- capture all messages from V8 and filter out all but the one for import assertion
- forward information about module that raised the warning and make it drop V8 code cache for this module

**Comment by bartlomieju** [comment]
Blocker for landing: this still only prints warning once, after that if V8 cache is used the warning will not be printed again.

**Comment by bartlomieju** [comment]
Blocked on https://github.com/denoland/deno_core/pull/873.

File: cli/module_loader.rs [hunk 1]
```diff
@@ -695,6 +700,10 @@ enum CodeOrDeferredEmit<'a> {
 // todo(dsherret): this double Rc boxing is not ideal
 struct CliModuleLoader<TGraphContainer: ModuleGraphContainer>(
   Rc<CliModuleLoaderInner<TGraphContainer>>,
+  // NOTE(bartlomieju): this is temporary, for deprecated import assertions.
+  // Should be removed in Deno 2.
+  // Modules stored here should not be V8 code-cached.
+  Arc<Mutex<HashSet<String>>>,
```
Comment (bartlomieju) [inline cli/module_loader.rs:706]: Move to `CliModuleLoaderInner`

File: cli/module_loader.rs [hunk 2]
```diff
@@ -342,6 +345,10 @@ struct CliModuleLoaderInner<TGraphContainer: ModuleGraphContainer> {
   emitter: Arc<Emitter>,
   parsed_source_cache: Arc<ParsedSourceCache>,
   graph_container: TGraphContainer,
+  // NOTE(bartlomieju): this is temporary, for deprecated import assertions.
+  // Should be removed in Deno 2.
+  // Modules stored here should not be V8 code-cached.
+  prevent_v8_code_cache: Arc<Mutex<HashSet<String>>>,
```
Comment (dsherret) [inline cli/module_loader.rs:351]: Does this need to be an in Arc? Maybe use `DashSet` instead?

**Review by dsherret — APPROVED** [review]
LGTM

File: cli/module_loader.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/module_loader.rs:351]: I don't think it's a big deal? This code is gonna be gone in a few days anyway.

**Pull request merged**